File: crates/fc-platform/src/application_openapi_spec/operations/diff.rs
//! Diff two OpenAPI documents into structured `ChangeNotes` plus a
//! human-readable summary.
//!
//! v1 is deliberately shallow: it set-diffs `paths` keys, `components.schemas`
//! keys, and the HTTP verbs of each surviving path. Anything removed is
//! flagged as `has_breaking`. Future versions can drill into request/response
//! schema shapes, but v1 surfaces the change-notes the user explicitly asked
//! for ("notes about what was removed") without pulling in a 3rd-party
//! OpenAPI parser.

use std::collections::BTreeSet;

use serde_json::Value;
use sha2::{Digest, Sha256};

use super::super::entity::ChangeNotes;

/// HTTP methods that appear as object keys under each path in OpenAPI.
const VERBS: &[&str] = &[
    "get", "put", "post", "delete", "options", "head", "patch", "trace",
];

/// Stable, canonical-JSON hash of a spec. Used for the no-op short-circuit so
/// re-sending the same document doesn't create a new row.
pub fn spec_hash(spec: &Value) -> String {
    let canonical = canonicalize(spec);
    let bytes = serde_json::to_vec(&canonical).unwrap_or_default();
    let mut hasher = Sha256::new();
    hasher.update(&bytes);
    let digest = hasher.finalize();
    digest.iter().fold(String::with_capacity(64), |mut acc, b| {
        use std::fmt::Write as _;
        let _ = write!(acc, "{:02x}", b);
        acc
    })
}

/// Recursively sort object keys so equivalent documents hash identically.
fn canonicalize(v: &Value) -> Value {
    match v {
        Value::Object(map) => {
            let mut sorted = serde_json::Map::new();
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            for k in keys {
                sorted.insert(k.clone(), canonicalize(&map[k]));
            }
            Value::Object(sorted)
        }
        Value::Array(items) => Value::Array(items.iter().map(canonicalize).collect()),
        other => other.clone(),
    }
}

/// Compute the structured diff and a pre-rendered human summary.
pub fn compute_change_notes(prior: &Value, current: &Value) -> (ChangeNotes, String) {
    let prior_paths = collect_object_keys(prior.pointer("/paths"));
    let current_paths = collect_object_keys(current.pointer("/paths"));
    let removed_paths: Vec<String> = prior_paths.difference(&current_paths).cloned().collect();
    let added_paths: Vec<String> = current_paths.difference(&prior_paths).cloned().collect();

    let prior_schemas = collect_object_keys(prior.pointer("/components/schemas"));
    let current_schemas = collect_object_keys(current.pointer("/components/schemas"));
    let removed_schemas: Vec<String> = prior_schemas
        .difference(&current_schemas)
        .cloned()
        .collect();
    let added_schemas: Vec<String> = current_schemas
        .difference(&prior_schemas)
        .cloned()
        .collect();

    // For paths present in both, compare verb sets to surface dropped operations.
    let mut removed_operations: Vec<String> = Vec::new();
    for path in prior_paths.intersection(&current_paths) {
        let prior_verbs = verbs_for_path(prior.pointer(&path_pointer(path)));
        let current_verbs = verbs_for_path(current.pointer(&path_pointer(path)));
        for v in prior_verbs.difference(&current_verbs) {
            removed_operations.push(format!("{} {}", v.to_uppercase(), path));
        }
    }

    let has_breaking =
        !removed_paths.is_empty() || !removed_schemas.is_empty() || !removed_operations.is_empty();

    let mut notes = ChangeNotes {
        added_paths,
        removed_paths,
        added_schemas,
        removed_schemas,
        removed_operations,
        has_breaking,
    };
    // Sort for deterministic output (sets aren't ordered).
    notes.added_paths.sort();
    notes.removed_paths.sort();
    notes.added_schemas.sort();
    notes.removed_schemas.sort();
    notes.removed_operations.sort();

    let summary = render_summary(&notes);
    (notes, summary)
}

fn collect_object_keys(node: Option<&Value>) -> BTreeSet<String> {
    match node {
        Some(Value::Object(map)) => map.keys().cloned().collect(),
        _ => BTreeSet::new(),
    }
}

fn verbs_for_path(node: Option<&Value>) -> BTreeSet<String> {
    let mut set = BTreeSet::new();
    if let Some(Value::Object(map)) = node {
        for v in VERBS {
            if map.contains_key(*v) {
                set.insert((*v).to_string());
            }
        }
    }
    set
}

/// Build a JSON pointer for `paths.<path>`. OpenAPI path strings contain `/`
/// and `~`, which are reserved in JSON pointers (RFC 6901), so escape them.
fn path_pointer(path: &str) -> String {
    let escaped = path.replace('~', "~0").replace('/', "~1");
    format!("/paths/{}", escaped)
}

fn render_summary(notes: &ChangeNotes) -> String {
    if notes.is_empty() {
        return "No structural changes (descriptions or examples may differ).".to_string();
    }
    let mut parts: Vec<String> = Vec::new();
    if !notes.added_paths.is_empty() {
        parts.push(format!("Added {} path(s)", notes.added_paths.len()));
    }
    if !notes.removed_paths.is_empty() {
        parts.push(format!("Removed {} path(s)", notes.removed_paths.len()));
    }
    if !notes.removed_operations.is_empty() {
        parts.push(format!(
            "Removed {} operation(s)",
            notes.removed_operations.len()
        ));
    }
    if !notes.added_schemas.is_empty() {
        parts.push(format!("Added {} schema(s)", notes.added_schemas.len()));
    }
    if !notes.removed_schemas.is_empty() {
        parts.push(format!(
            "Removed {} schema(s)",
            notes.removed_schemas.len()
        ));
    }

    let mut summary = parts.join("; ");
    if notes.has_breaking {
        summary.push_str(". Contains breaking changes (removals).");
    } else {
        summary.push('.');
    }

    // Append a short tail of the actual identifiers so the listing tells you
    // what changed at a glance.
    let mut details: Vec<String> = Vec::new();
    if !notes.removed_paths.is_empty() {
        details.push(format!(
            "removed paths: {}",
            sample(&notes.removed_paths, 5)
        ));
    }
    if !notes.removed_operations.is_empty() {
        details.push(format!(
            "removed ops: {}",
            sample(&notes.removed_operations, 5)
        ));
    }
    if !notes.removed_schemas.is_empty() {
        details.push(format!(
            "removed schemas: {}",
            sample(&notes.removed_schemas, 5)
        ));
    }
    if !details.is_empty() {
        summary.push_str(" — ");
        summary.push_str(&details.join("; "));
    }
    summary
}

fn sample(items: &[String], n: usize) -> String {
    if items.len() <= n {
        items.join(", ")
    } else {
        format!("{}, … (+{} more)", items[..n].join(", "), items.len() - n)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn hash_is_stable_across_key_order() {
        let a = json!({"a": 1, "b": {"x": 1, "y": 2}});
        let b = json!({"b": {"y": 2, "x": 1}, "a": 1});
        assert_eq!(spec_hash(&a), spec_hash(&b));
    }

    #[test]
    fn diff_finds_added_and_removed_paths() {
        let prior = json!({"paths": {"/a": {"get": {}}, "/b": {"get": {}}}});
        let current = json!({"paths": {"/a": {"get": {}}, "/c": {"post": {}}}});
        let (notes, summary) = compute_change_notes(&prior, &current);
        assert_eq!(notes.removed_paths, vec!["/b"]);
        assert_eq!(notes.added_paths, vec!["/c"]);
        assert!(notes.has_breaking);
        assert!(summary.contains("breaking"));
    }

    #[test]
    fn diff_finds_removed_verbs() {
        let prior = json!({"paths": {"/a": {"get": {}, "post": {}}}});
        let current = json!({"paths": {"/a": {"get": {}}}});
        let (notes, _) = compute_change_notes(&prior, &current);
        assert_eq!(notes.removed_operations, vec!["POST /a"]);
        assert!(notes.has_breaking);
    }

    #[test]
    fn diff_no_changes_is_not_breaking() {
        let spec = json!({"paths": {"/a": {"get": {}}}});
        let (notes, summary) = compute_change_notes(&spec, &spec);
        assert!(!notes.has_breaking);
        assert!(notes.is_empty());
        assert!(summary.starts_with("No structural"));
    }

    #[test]
    fn diff_finds_schema_changes() {
        let prior = json!({"components": {"schemas": {"User": {}, "Order": {}}}});
        let current = json!({"components": {"schemas": {"User": {}, "Invoice": {}}}});
        let (notes, _) = compute_change_notes(&prior, &current);
        assert_eq!(notes.removed_schemas, vec!["Order"]);
        assert_eq!(notes.added_schemas, vec!["Invoice"]);
    }

    #[test]
    fn path_pointer_escapes_slashes() {
        assert_eq!(path_pointer("/users/{id}"), "/paths/~1users~1{id}");
    }
}

File: internal/platform/openapispecs/diff.go
package openapispecs

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// httpVerbs is the set of OpenAPI path-item keys we treat as
// "operations" — anything else (parameters, summary, $ref) isn't a
// verb so it doesn't count toward removed-operation breakage.
var httpVerbs = []string{
	"get", "put", "post", "delete", "options", "head", "patch", "trace",
}

// SpecHash returns a stable canonical-JSON sha256 of the spec. Used by
// the no-op short-circuit so re-sending the same document doesn't
// create a new row.
func SpecHash(spec json.RawMessage) string {
	var parsed any
	if err := json.Unmarshal(spec, &parsed); err != nil {
		return ""
	}
	canonical, err := canonicalJSON(parsed)
	if err != nil {
		return ""
	}
	sum := sha256.Sum256(canonical)
	return hex.EncodeToString(sum[:])
}

// canonicalJSON serialises v with object keys in sorted order so two
// semantically-equal documents hash identically.
func canonicalJSON(v any) ([]byte, error) {
	var b strings.Builder
	if err := writeCanonical(&b, v); err != nil {
		return nil, err
	}
	return []byte(b.String()), nil
}

func writeCanonical(b *strings.Builder, v any) error {
	switch t := v.(type) {
	case map[string]any:
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		b.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				b.WriteByte(',')
			}
			kj, err := json.Marshal(k)
			if err != nil {
				return err
			}
			b.Write(kj)
			b.WriteByte(':')
			if err := writeCanonical(b, t[k]); err != nil {
				return err
			}
		}
		b.WriteByte('}')
	case []any:
		b.WriteByte('[')
		for i, item := range t {
			if i > 0 {
				b.WriteByte(',')
			}
			if err := writeCanonical(b, item); err != nil {
				return err
			}
		}
		b.WriteByte(']')
	default:
		raw, err := json.Marshal(v)
		if err != nil {
			return err
		}
		b.Write(raw)
	}
	return nil
}

// ComputeChangeNotes diffs two OpenAPI documents and returns the
// structured diff plus a human-readable summary suitable for listings.
// Mirrors the Rust shallow-diff: set-diff `paths` keys + `components.schemas`
// keys + verbs of surviving paths. Anything removed → has_breaking.
func ComputeChangeNotes(prior, current json.RawMessage) (ChangeNotes, string) {
	var priorDoc, currentDoc map[string]any
	_ = json.Unmarshal(prior, &priorDoc)
	_ = json.Unmarshal(current, &currentDoc)

	priorPaths := objectKeys(nestedObject(priorDoc, "paths"))
	currentPaths := objectKeys(nestedObject(currentDoc, "paths"))
	priorSchemas := objectKeys(nestedObject(priorDoc, "components", "schemas"))
	currentSchemas := objectKeys(nestedObject(currentDoc, "components", "schemas"))

	notes := ChangeNotes{
		AddedPaths:     setDifference(currentPaths, priorPaths),
		RemovedPaths:   setDifference(priorPaths, currentPaths),
		AddedSchemas:   setDifference(currentSchemas, priorSchemas),
		RemovedSchemas: setDifference(priorSchemas, currentSchemas),
	}

	// Verbs dropped from paths that survive.
	survivors := setIntersection(priorPaths, currentPaths)
	for _, path := range survivors {
		priorVerbs := verbsForPath(nestedObject(priorDoc, "paths", path))
		currentVerbs := verbsForPath(nestedObject(currentDoc, "paths", path))
		for _, v := range priorVerbs {
			if !contains(currentVerbs, v) {
				notes.RemovedOperations = append(notes.RemovedOperations,
					strings.ToUpper(v)+" "+path)
			}
		}
	}
	sort.Strings(notes.AddedPaths)
	sort.Strings(notes.RemovedPaths)
	sort.Strings(notes.AddedSchemas)
	sort.Strings(notes.RemovedSchemas)
	sort.Strings(notes.RemovedOperations)

	notes.HasBreaking = len(notes.RemovedPaths) > 0 ||
		len(notes.RemovedSchemas) > 0 ||
		len(notes.RemovedOperations) > 0

	return notes, renderSummary(notes)
}

// ── helpers ──────────────────────────────────────────────────────────────

func nestedObject(root map[string]any, path ...string) map[string]any {
	cur := root
	for _, p := range path {
		if cur == nil {
			return nil
		}
		next, ok := cur[p].(map[string]any)
		if !ok {
			return nil
		}
		cur = next
	}
	return cur
}

func objectKeys(o map[string]any) []string {
	if o == nil {
		return nil
	}
	out := make([]string, 0, len(o))
	for k := range o {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

func verbsForPath(node map[string]any) []string {
	var out []string
	for _, v := range httpVerbs {
		if _, ok := node[v]; ok {
			out = append(out, v)
		}
	}
	return out
}

func setDifference(a, b []string) []string {
	bset := make(map[string]struct{}, len(b))
	for _, x := range b {
		bset[x] = struct{}{}
	}
	out := []string{}
	for _, x := range a {
		if _, in := bset[x]; !in {
			out = append(out, x)
		}
	}
	return out
}

func setIntersection(a, b []string) []string {
	bset := make(map[string]struct{}, len(b))
	for _, x := range b {
		bset[x] = struct{}{}
	}
	out := []string{}
	for _, x := range a {
		if _, in := bset[x]; in {
			out = append(out, x)
		}
	}
	return out
}

func contains(haystack []string, needle string) bool {
	for _, x := range haystack {
		if x == needle {
			return true
		}
	}
	return false
}

func renderSummary(n ChangeNotes) string {
	if n.IsEmpty() {
		return "No structural changes (descriptions or examples may differ)."
	}
	var parts []string
	if len(n.AddedPaths) > 0 {
		parts = append(parts, fmt.Sprintf("Added %d path(s)", len(n.AddedPaths)))
	}
	if len(n.RemovedPaths) > 0 {
		parts = append(parts, fmt.Sprintf("Removed %d path(s)", len(n.RemovedPaths)))
	}
	if len(n.RemovedOperations) > 0 {
		parts = append(parts, fmt.Sprintf("Removed %d operation(s)", len(n.RemovedOperations)))
	}
	if len(n.AddedSchemas) > 0 {
		parts = append(parts, fmt.Sprintf("Added %d schema(s)", len(n.AddedSchemas)))
	}
	if len(n.RemovedSchemas) > 0 {
		parts = append(parts, fmt.Sprintf("Removed %d schema(s)", len(n.RemovedSchemas)))
	}
	summary := strings.Join(parts, "; ")
	if n.HasBreaking {
		summary += ". Contains breaking changes (removals)."
	} else {
		summary += "."
	}
	var details []string
	if len(n.RemovedPaths) > 0 {
		details = append(details, "removed paths: "+sample(n.RemovedPaths, 5))
	}
	if len(n.RemovedOperations) > 0 {
		details = append(details, "removed ops: "+sample(n.RemovedOperations, 5))
	}
	if len(n.RemovedSchemas) > 0 {
		details = append(details, "removed schemas: "+sample(n.RemovedSchemas, 5))
	}
	if len(details) > 0 {
		summary += " — " + strings.Join(details, "; ")
	}
	return summary
}

func sample(items []string, n int) string {
	if len(items) <= n {
		return strings.Join(items, ", ")
	}
	return fmt.Sprintf("%s, … (+%d more)", strings.Join(items[:n], ", "), len(items)-n)
}
